File: src/odioctl/web/services.py
"""What the pages can ask the box to do — no HTTP, no HTML.

One operation per form: toggle a component, run a component's own command,
start an upgrade, select a DAC. state.json is edited in this process (it runs
as the odios target user); only config.txt writes escalate, through
`sudo -n odioctl dac set <id>`. Upgrades are never run here — the web process
starts `odio-upgrade.service`, the unit odio-api drives too.

Every subprocess goes through an injected callable (`privileged_run`,
`user_run`, `action_spawn`), so the tests drive the real code path against
stand-ins instead of the box.
"""

from __future__ import annotations

import json
import os
import secrets
import subprocess
import threading
from collections import deque
from collections.abc import Callable
from dataclasses import dataclass
from typing import Any, cast

from odioctl import components, dac, state
from odioctl.upgrade import check
from odioctl.web.config import UPGRADE_UNIT, WebConfig

RunFn = Callable[[list[str]], "subprocess.CompletedProcess[str]"]
SpawnFn = Callable[[list[str]], "subprocess.Popen[str]"]

# How long a component action gets to print its link before we give up on it.
# `qbzd login` fetches an app id over the network first, so it is not instant.
ACTION_LINK_TIMEOUT = 15.0


class WebError(Exception):
    """User-facing failure of a form action; rendered as an error banner."""


class TokenError(WebError):
    """Missing or invalid form token; answered with a bare 403, not a re-render."""


def _default_privileged_run(cfg: WebConfig) -> RunFn:
    def run(args: list[str]) -> subprocess.CompletedProcess[str]:
        return subprocess.run(
            ["sudo", "-n", cfg.odioctl_bin, *args],
            capture_output=True,
            text=True,
            timeout=30,
            check=False,
        )

    return run


def _default_user_run(args: list[str]) -> subprocess.CompletedProcess[str]:
    return subprocess.run(args, capture_output=True, text=True, timeout=30, check=False)


def _run_checked(run: RunFn, args: list[str], what: str) -> None:
    """Run a subprocess through `run`, turning any failure into a WebError banner."""
    try:
        proc = run(args)
    except (OSError, subprocess.SubprocessError) as e:
        raise WebError(f"cannot run {what}: {e}") from e
    if proc.returncode != 0:
        detail = (proc.stderr or proc.stdout or "").strip() or f"exit {proc.returncode}"
        raise WebError(f"{what} failed: {detail}")


def _default_action_spawn(argv: list[str]) -> subprocess.Popen[str]:
    # Same user as this process (the odios target user), no sudo: a component
    # action is exactly what the operator would type on the box.
    return subprocess.Popen(
        argv,
        stdout=subprocess.PIPE,
        stderr=subprocess.STDOUT,
        stdin=subprocess.DEVNULL,
        text=True,
        bufsize=1,
    )


def _read_until_link(run: ActionRun, scheme: str, timeout: float) -> None:
    """Drain the process output until it prints a link, EOF, or `timeout`.

    Keeps draining in the background afterwards: the child writes a few more
    lines when the user follows the link, and a full pipe would wedge it. It
    stops *recording* there, though — what a login helper prints once the user
    is through is the credential it just obtained (upmpdcli's Tidal helper
    dumps the access and refresh tokens), and nothing that lands in `output`
    is worth painting into a browser.
    """
    found = threading.Event()

    def drain() -> None:
        assert run.proc.stdout is not None
        for line in run.proc.stdout:
            if run.url:
                continue
            run.output.append(line)
            token = next((w for w in line.split() if w.startswith(scheme)), "")
            if token:
                run.url = token
                found.set()
        found.set()  # EOF: nothing more is coming

    threading.Thread(target=drain, daemon=True).start()
    found.wait(timeout)


@dataclass
class ActionResult:
    """What an action just did, shown once in the modal of the POST response.

    Popped by `render_page`, so it never survives into the next page load:
    what outlives the request is the pending link on the component row.
    """

    title: str
    output: str
    url: str = ""
    link_label: str = ""


@dataclass
class ActionRun:
    """One started action: its process, its output, and the link it printed."""

    proc: subprocess.Popen[str]
    output: deque[str]
    url: str = ""

    def alive(self) -> bool:
        return self.proc.poll() is None


class Services:
    """Business operations behind the pages (also unit-testable directly)."""

    def __init__(
        self,
        cfg: WebConfig,
        privileged_run: RunFn | None = None,
        user_run: RunFn | None = None,
        action_spawn: SpawnFn | None = None,
    ) -> None:
        self.cfg = cfg
        self.privileged_run = privileged_run or _default_privileged_run(cfg)
        self.user_run = user_run or _default_user_run  # same user, no sudo (systemctl --user)
        self.action_spawn = action_spawn or _default_action_spawn
        self.token = secrets.token_urlsafe(24)
        self._lock = threading.Lock()
        # Started actions outlive their request: `qbzd login` waits up to 300s
        # for the user to follow its link. Keyed by (kind, name, action id).
        self._runs: dict[tuple[str, str, str], ActionRun] = {}
        self._notes: dict[tuple[str, str, str], str] = {}
        self._result: ActionResult | None = None

    # -- reads ----------------------------------------------------------------

    def read_state(self) -> tuple[state.State | None, str | None]:
        """(state, None) or (None, error message)."""
        try:
            return state.read_state(self.cfg.state_path), None
        except FileNotFoundError:
            return None, f"{self.cfg.state_path} not found"
        except (OSError, json.JSONDecodeError, state.StateError) as e:
            return None, f"cannot read state.json: {e}"

    def dac_status(self) -> dict[str, Any]:
        return dac.status(self.cfg.config_txt)

    def upgrade_report(self) -> check.UpgradeReport | None:
        return check.read_report(self.cfg.resolved_upgrades_path())

    # -- writes ---------------------------------------------------------------

    def set_component(self, kind: str, name: str, enabled: bool) -> str:
        if kind not in ("role", "feature"):
            raise WebError(f"unknown component kind {kind!r}")
        with self._lock:
            st, err = self.read_state()
            if st is None:
                raise WebError(err or "state.json unavailable")
            try:
                new = components.set_component(st, cast(components.Kind, kind), name, enabled)
            except components.ComponentError as e:
                raise WebError(str(e)) from e
            try:
                components.save(self.cfg.state_path, new)
            except OSError as e:
                raise WebError(f"cannot write state.json: {e}") from e
            # Keep upgrades.json in step so odio-ui's badge and `upgrade apply`
            # see the pending install without waiting for the daily timer.
            report = check.refresh(
                check.CheckOptions(
                    state=self.cfg.state_path, output=self.cfg.resolved_upgrades_path()
                )
            )
        label = components.label_of(cast(components.Kind, kind), name)
        if not enabled:
            return f"{label} disabled — it stays installed but will no longer be updated."
        if report is not None and f"{kind}:{name}" in report["pending_components"]:
            return f"{label} enabled — it will be installed by the next upgrade (apply it below)."
        return f"{label} enabled."

    def run_action(self, kind: str, name: str, action_id: str, host: str) -> str:
        """Start a catalog action and return the banner text for the page.

        The command is never waited on: `qbzd login` prints its Qobuz URL and
        then holds a listener open until the browser comes back (300s), so we
        read stdout only until the link shows up and leave the process to it.
        The link is then rendered next to the component until it is gone.
        """
        if kind not in ("role", "feature"):
            raise WebError(f"unknown component kind {kind!r}")
        action = components.find_action(cast(components.Kind, kind), name, action_id)
        if action is None:
            raise WebError(f"unknown action {action_id!r} for {name}")
        st, err = self.read_state()
        if st is None:
            raise WebError(err or "state.json unavailable")
        comp = next(
            (c for c in components.list_components(st) if c.kind == kind and c.name == name), None
        )
        if comp is None or comp.status != "installed":
            raise WebError(
                f"{components.label_of(cast(components.Kind, kind), name)} is not installed"
            )

        key = (kind, name, action_id)
        with self._lock:
            run = self._runs.get(key)
            if run is not None and run.alive():
                self._result = ActionResult(
                    title=action.label,
                    output="".join(run.output),
                    url=run.url,
                    link_label=action.link_label,
                )
                return f"{action.label}: already running — the link is below."
            self._notes.pop(key, None)
            argv = [part.format(host=host, home=os.path.expanduser("~")) for part in action.argv]
            try:
                proc = self.action_spawn(argv)
            except (OSError, subprocess.SubprocessError) as e:
                raise WebError(f"cannot run {' '.join(argv)}: {e}") from e
            run = ActionRun(proc=proc, output=deque(maxlen=20))
            self._runs[key] = run
        _read_until_link(run, action.link_scheme, ACTION_LINK_TIMEOUT)
        output = "".join(run.output)
        with self._lock:
            self._result = ActionResult(
                title=action.label, output=output, url=run.url, link_label=action.link_label
            )
        if run.url:
            return f"{action.label}: open the link below to finish."
        with self._lock:
            del self._runs[key]
        # No link: either it died (reap it for the exit code — stdout can close
        # a moment before the process does) or it is stuck and we stop it.
        try:
            run.proc.wait(timeout=2)
        except subprocess.TimeoutExpired:
            run.proc.terminate()
        if run.proc.returncode is None:
            raise WebError(f"{action.label}: no link after {ACTION_LINK_TIMEOUT:.0f}s")
        raise WebError(f"{action.label} failed (exit {run.proc.returncode})")

    def action_state(self, kind: str, name: str, action_id: str) -> tuple[str, str]:
        """(pending link, note) for one action — ("", "") when it never ran.

        Reaps a finished run, turning it into the note the page shows on the
        next render: no JavaScript here, the operator reloads to see the end.
        """
        key = (kind, name, action_id)
        with self._lock:
            run = self._runs.get(key)
            if run is None:
                return "", self._notes.get(key, "")
            if run.alive():
                return run.url, ""
            del self._runs[key]
            if run.proc.returncode == 0:
                note = "Done."
            else:
                detail = " ".join(line.strip() for line in run.output if line.strip())[-200:]
                note = f"Failed (exit {run.proc.returncode}). {detail}".strip()
            self._notes[key] = note
            return "", note

    def pop_action_result(self) -> ActionResult | None:
        """The last action's output, once: the modal shows it and it is gone."""
        with self._lock:
            result, self._result = self._result, None
            return result

    def start_upgrade(self) -> str:
        """Start the odio-upgrade user unit (= `sudo odioctl upgrade apply --progress`)."""
        report = self.upgrade_report()
        if report is None or not report["upgrade_available"]:
            raise WebError("nothing to apply — no upgrade or pending component reported")
        _run_checked(
            self.user_run,
            ["systemctl", "--user", "start", "--no-block", UPGRADE_UNIT],
            f"systemctl --user start {UPGRADE_UNIT}",
        )
        return "Upgrade started — follow its progress in odio-ui."

    def set_dac(self, dac_id: str | None) -> str:
        if dac_id is not None and dac_id not in dac.BY_ID:
            raise WebError(f"unknown DAC id {dac_id!r}")
        args = ["dac", "unset"] if dac_id is None else ["dac", "set", dac_id]
        if self.cfg.config_txt:
            args += ["--config", self.cfg.config_txt]
        _run_checked(self.privileged_run, args, "odioctl dac")
        what = "DAC block removed" if dac_id is None else f"DAC set to {dac_id}"
        return f"{what} — reboot required."

File: web/services.go
package web

// What the pages can ask the box to do — no HTTP, no HTML. One operation per
// form. state.json is edited in this process; only config.txt writes escalate
// (`sudo -n odioctl dac …`). Upgrades are never run here: the web process
// starts odio-upgrade.service, the unit odio-api drives too, and watches it
// through systemd. Subprocesses go through Runners so tests drive the real
// code path against stand-ins.

import (
	"crypto/rand"
	"encoding/base64"
	"fmt"
	"io"
	"log"
	"os"
	"strconv"
	"strings"
	"sync"
	"time"

	"github.com/b0bbywan/odioctl/components"
	"github.com/b0bbywan/odioctl/dac"
	"github.com/b0bbywan/odioctl/state"
	"github.com/b0bbywan/odioctl/upgrade"
)

// actionLinkTimeout is how long a component action gets to print its link.
// `qbzd login` fetches an app id over the network first, so it is not instant.
var actionLinkTimeout = 15 * time.Second

// upgradePoll is how often the watcher asks systemd about odio-upgrade.service
// when it cannot be told (no units directory to watch); upgradeRecheck is
// the net behind the events when it can.
var (
	upgradePoll    = 2 * time.Second
	upgradeRecheck = 15 * time.Second
)

// UserError is a failure the page shows as an error banner; Modal carries the
// action's output when there is some to show alongside it.
type UserError struct {
	Msg   string
	Modal *ActionResult
}

func (e *UserError) Error() string { return e.Msg }

func userErrorf(format string, args ...any) error {
	return &UserError{Msg: fmt.Sprintf(format, args...)}
}

type RunResult struct {
	Stdout, Stderr string
	Code           int
}

type RunFn func(args []string) (RunResult, error)

// ActionProcess is one started component action, as the services see it.
type ActionProcess interface {
	Output() io.Reader // combined stdout+stderr
	Pid() int
	Alive() bool
	ExitCode() int                // valid once !Alive()
	WaitFor(d time.Duration) bool // true when the process exited within d
	Stop()                        // terminate, then kill — nothing polls it again
}

// Runners are the subprocess seams; NewServices fills zero fields with the
// real ones.
type Runners struct {
	Privileged RunFn // sudo -n odioctl …
	User       RunFn // same user, no sudo (systemctl --user)
	Spawn      func(argv []string) (ActionProcess, error)
	Log        io.Writer // what the process did, one line each; nil → stderr
}

// ActionResult is what an action just did, shown in the modal of the POST
// response — it never reaches another client or the next page load; what
// outlives the request is the row's own link.
type ActionResult struct {
	ID                            string // element id, so a later fragment replaces this modal
	Title, Output, URL, LinkLabel string
	Note                          actionNote // how it ended, once it has: a Done button where the link was
}

type actionKey struct {
	kind components.Kind
	name string
	id   string
}

// domID names the modal of this action in the page, so a fragment for its
// end replaces that one and no other.
func (k actionKey) domID() string { return fmt.Sprintf("modal-%s-%s-%s", k.kind, k.name, k.id) }

// Services holds the business operations behind the pages (also
// unit-testable directly).
type Services struct {
	cfg   Config
	run   Runners
	token string
	log   *log.Logger

	mu sync.Mutex
	// Started actions outlive their request: `qbzd login` waits up to 300s
	// for the user to follow its link.
	runs map[actionKey]*actionRun
	// Reaped runs, kept for their note (the row) and their output (the
	// modal that showed their link gets its end the same way).
	finished map[actionKey]*actionRun
	// A watcher polls odio-upgrade.service to its end; how the last one ended.
	watching    bool
	upgradeNote actionNote
	// Closed when something the page shows has changed, then replaced: an
	// /events stream waits on it and takes the next one.
	wake chan struct{}
}

func NewServices(cfg Config, r Runners) *Services {
	if r.Privileged == nil {
		r.Privileged = defaultPrivilegedRun(cfg)
	}
	if r.User == nil {
		r.User = defaultUserRun
	}
	if r.Spawn == nil {
		r.Spawn = defaultSpawn
	}
	if r.Log == nil {
		r.Log = os.Stderr
	}
	return &Services{
		cfg:   cfg,
		run:   r,
		token: newToken(),
		// journald stamps the lines itself; under `go run` the shell does.
		log:      log.New(r.Log, "", 0),
		runs:     map[actionKey]*actionRun{},
		finished: map[actionKey]*actionRun{},
		wake:     make(chan struct{}),
	}
}

// Wake is the channel closed on the next change; take it before reading
// what it guards, and a fresh one once it has fired.
func (s *Services) Wake() <-chan struct{} {
	s.mu.Lock()
	defer s.mu.Unlock()
	return s.wake
}

// changed wakes every waiter at once: close, then a new channel for the next.
func (s *Services) changed() {
	s.mu.Lock()
	defer s.mu.Unlock()
	close(s.wake)
	s.wake = make(chan struct{})
}

func newToken() string {
	b := make([]byte, 24)
	if _, err := rand.Read(b); err != nil {
		panic(err)
	}
	return base64.RawURLEncoding.EncodeToString(b)
}

// Token is the per-process form token every POST must echo.
func (s *Services) Token() string { return s.token }

func (s *Services) Config() Config { return s.cfg }

// -- reads --------------------------------------------------------------

func (s *Services) ReadState() (state.State, error) {
	return state.Read(s.cfg.StatePath)
}

func (s *Services) DacStatus() dac.Status {
	return dac.GetStatus(s.cfg.ConfigTxt)
}

func (s *Services) UpgradeReport() *upgrade.Report {
	return upgrade.ReadReport(s.cfg.ResolvedUpgradesPath())
}

// AvailableRoles is the target release's role set, nil until a check has run.
func (s *Services) AvailableRoles() map[string]string {
	if report := s.UpgradeReport(); report != nil {
		return report.Manifest.Roles
	}
	return nil
}

// -- writes -------------------------------------------------------------

func (s *Services) SetComponent(kind components.Kind, name string, enabled bool) (string, error) {
	if err := s.writeComponent(kind, name, enabled); err != nil {
		return "", err
	}
	// Keep upgrades.json in step so odio-ui's badge and `upgrade apply` see
	// the pending install without waiting for the daily timer. Outside the
	// lock: it fetches the manifest, and every render takes that lock.
	report := upgrade.Refresh(upgrade.CheckOptions{
		State:  s.cfg.StatePath,
		Output: s.cfg.ResolvedUpgradesPath(),
	})
	s.changed()
	label := components.LabelOf(kind, name)
	switch {
	case !enabled:
		return label + " disabled — it stays installed but will no longer be updated.", nil
	case report != nil && report.HasPending(string(kind)+":"+name):
		return label + " enabled — it will be installed by the next upgrade (apply it below).", nil
	default:
		return label + " enabled.", nil
	}
}

func (s *Services) writeComponent(kind components.Kind, name string, enabled bool) error {
	s.mu.Lock()
	defer s.mu.Unlock()
	st, err := s.ReadState()
	if err != nil {
		return &UserError{Msg: stateErrorMsg(s.cfg.StatePath, err)}
	}
	next, err := components.Set(st, kind, name, enabled)
	if err != nil {
		return &UserError{Msg: err.Error()}
	}
	if err := state.Write(s.cfg.StatePath, next); err != nil {
		return userErrorf("cannot write state.json: %v", err)
	}
	return nil
}

// resolveAction is the catalog action to run, or a *UserError naming what is
// wrong. Offered for installed components only.
func (s *Services) resolveAction(kind components.Kind, name, id string) (components.Action, error) {
	action, ok := components.FindAction(kind, name, id)
	if !ok {
		return action, userErrorf("unknown action %q for %s", id, name)
	}
	st, err := s.ReadState()
	if err != nil {
		return action, &UserError{Msg: stateErrorMsg(s.cfg.StatePath, err)}
	}
	for _, c := range components.List(st, nil) {
		if c.Kind == kind && c.Name == name && c.Status == components.Installed {
			return action, nil
		}
	}
	return action, userErrorf("%s is not installed", components.LabelOf(kind, name))
}

// RunAction starts a catalog action and returns (banner, modal). The command
// is never waited on: `qbzd login` prints its URL and then holds a listener
// open until the browser comes back (300s), so stdout is read only until the
// link shows up and the process is left to it.
func (s *Services) RunAction(kind components.Kind, name, id, host string) (string, *ActionResult, error) {
	action, err := s.resolveAction(kind, name, id)
	if err != nil {
		return "", nil, err
	}
	key := actionKey{kind, name, id}

	s.mu.Lock()
	if run, ok := s.runs[key]; ok && run.alive() {
		s.mu.Unlock()
		s.log.Printf("action %s/%s: already running (pid %d), showing its link again", name, id, run.proc.Pid())
		return action.Label + ": already running — the link is below.", run.result(), nil
	}
	delete(s.finished, key)
	run, err := startAction(s.run.Spawn, action, host, s.cfg.Home)
	if err != nil {
		s.mu.Unlock()
		return "", nil, err
	}
	run.id = key.domID()
	s.runs[key] = run
	s.mu.Unlock()
	s.log.Printf("action %s/%s: spawned pid %d: %s", name, id, run.proc.Pid(), strings.Join(run.argv, " "))
	go func() { // the exit, when it happens: logged, and the open pages told
		code := run.proc.ExitCode()
		s.log.Printf("action %s/%s: pid %d exited %d after %s", name, id, run.proc.Pid(), code, run.elapsed())
		s.changed()
	}()

	if url := run.awaitLink(actionLinkTimeout); url != "" {
		s.log.Printf("action %s/%s: link after %s: %s", name, id, run.elapsed(), url)
		s.changed() // the row shows the link while the process lives
		return action.Label + ": open the link below to finish.", run.result(), nil
	}
	s.log.Printf("action %s/%s: no link after %s, output so far: %q", name, id, run.elapsed(), run.text())

	// No link: either it died (reap it for the exit code — stdout can close a
	// moment before the process does) or it is stuck and we stop it. The
	// entry stays in s.runs meanwhile: it is what keeps a second click from
	// spawning again while this one is still being stopped, and what
	// ActionState turns into the row's "Failed (exit N)" note on reload.
	if !run.proc.WaitFor(2 * time.Second) {
		run.proc.Stop()
		return "", nil, &UserError{
			Msg:   fmt.Sprintf("%s: no link after %.0fs", action.Label, actionLinkTimeout.Seconds()),
			Modal: run.result(),
		}
	}
	return "", nil, &UserError{
		Msg:   fmt.Sprintf("%s failed (exit %d)", action.Label, run.proc.ExitCode()),
		Modal: run.result(),
	}
}

// ActionState is the (pending link, note) of one action — ("", "") when it
// never ran. Reaps a finished run into the note the next render shows.
func (s *Services) ActionState(kind components.Kind, name, id string) (url string, note actionNote) {
	key := actionKey{kind, name, id}
	s.mu.Lock()
	defer s.mu.Unlock()
	s.reapLocked()
	if run, ok := s.runs[key]; ok {
		return run.link(), actionNote{}
	}
	if done := s.finished[key]; done != nil {
		return "", done.note()
	}
	return "", actionNote{}
}

// reapLocked moves every exited run to finished; s.mu held. A run being
// stopped is still alive and stays where a second click finds it.
func (s *Services) reapLocked() {
	for key, run := range s.runs {
		if !run.alive() {
			delete(s.runs, key)
			s.finished[key] = run
		}
	}
}

// FinishedResults is the end of every exited action, for the modal that
// showed its link: the same output, a Done button where the link was.
func (s *Services) FinishedResults() []*ActionResult {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.reapLocked()
	var out []*ActionResult
	for _, run := range s.finished {
		out = append(out, run.result())
	}
	return out
}

// upgrading reports whether a watcher follows odio-upgrade.service.
func (s *Services) upgrading() bool {
	s.mu.Lock()
	defer s.mu.Unlock()
	return s.watching
}

// StartUpgrade starts odio-upgrade.service (= `sudo odioctl upgrade apply
// --progress`) and watches it. Only this token-checked POST ever starts the
// unit: a render observes. A refusal inside the unit (sudoers) fails it
// within a second, and the watcher reports that like any other end.
func (s *Services) StartUpgrade() (string, error) {
	report := s.UpgradeReport()
	if report == nil || !report.UpgradeAvailable {
		return "", userErrorf("nothing to apply — no upgrade or pending component reported")
	}
	if s.upgrading() {
		return "Upgrade already running.", nil
	}
	args := []string{"systemctl", "--user", "start", "--no-block", UpgradeUnit}
	if err := runChecked(s.run.User, args, "systemctl --user start "+UpgradeUnit); err != nil {
		return "", err
	}
	s.log.Printf("upgrade: started %s", UpgradeUnit)
	s.watchUpgrade()
	s.changed()
	return "Upgrade started.", nil
}

// unitState is what `systemctl show` says of odio-upgrade.service.
type unitState struct {
	Active string // activating while the oneshot runs, then inactive or failed
	Result string // success, exit-code, …
	Code   int    // ExecMainStatus
}

func (u unitState) running() bool { return u.Active == "activating" }

// note is the end of a run, as the card shows it; a unit never run or run
// to success reads the same (inactive, success).
func (u unitState) note() actionNote {
	if u.Active == "failed" {
		return actionNote{Text: fmt.Sprintf("Failed (%s, exit %d).", u.Result, u.Code), Failed: true}
	}
	return actionNote{Text: "Done."}
}

// showUpgrade asks systemd; the error is logged, and read as not running.
// `show` prints Key=Value lines in systemd's own order, not the -p order
// (Result and ExecMainStatus come before ActiveState), hence keyed.
func (s *Services) showUpgrade() (unitState, error) {
	res, err := s.run.User([]string{"systemctl", "--user", "show",
		"-p", "ActiveState", "-p", "Result", "-p", "ExecMainStatus", UpgradeUnit})
	if err == nil && res.Code != 0 {
		err = fmt.Errorf("exit %d: %s", res.Code, strings.TrimSpace(res.Stderr))
	}
	if err != nil {
		s.log.Printf("upgrade: systemctl show %s: %v", UpgradeUnit, err)
		return unitState{}, err
	}
	var u unitState
	for _, line := range strings.Split(res.Stdout, "\n") {
		key, value, _ := strings.Cut(strings.TrimSpace(line), "=")
		switch key {
		case "ActiveState":
			u.Active = value
		case "Result":
			u.Result = value
		case "ExecMainStatus":
			u.Code, _ = strconv.Atoi(value)
		}
	}
	return u, nil
}

// watchUpgrade follows the unit to its end, then keeps that as the note and
// wakes the open pages. One watcher at a time.
func (s *Services) watchUpgrade() {
	s.mu.Lock()
	defer s.mu.Unlock()
	if s.watching {
		return
	}
	s.watching = true
	s.upgradeNote = actionNote{}
	go s.followUpgrade(time.Now())
}

// followUpgrade waits to be told the unit is over, then asks systemd how.
// The word comes from its invocation link leaving the units directory
// (systemd removes it as the unit leaves activating) with a slow tick
// behind; without the directory it is polled. No probe up front: `start
// --no-block` returns before the unit is activating.
func (s *Services) followUpgrade(started time.Time) {
	interval := upgradeRecheck
	gone, stop, err := removedFrom(s.cfg.UnitsDir(), "invocation:"+UpgradeUnit, s.log.Printf)
	if err != nil {
		s.log.Printf("upgrade: cannot watch %s (%v), polling %s", s.cfg.UnitsDir(), err, UpgradeUnit)
		interval = upgradePoll // gone is nil: never fires
	} else {
		defer stop()
	}
	tick := time.NewTicker(interval)
	defer tick.Stop()
	for {
		select {
		case <-gone:
		case <-tick.C:
		}
		if u, err := s.showUpgrade(); err == nil && !u.running() {
			s.endUpgrade(u, started)
			return
		}
	}
}

// endUpgrade keeps how the unit ended as the note and wakes the pages.
func (s *Services) endUpgrade(u unitState, started time.Time) {
	note := u.note()
	s.log.Printf("upgrade: %s %s after %s (%s)", UpgradeUnit, u.Active, time.Since(started).Round(time.Second), note.Text)
	s.mu.Lock()
	s.watching, s.upgradeNote = false, note
	s.mu.Unlock()
	s.changed()
}

// UpgradeState is (running, note of the last run). Without a watcher of its
// own, and while the report says there is something to apply, it asks
// systemd: a unit activating was started by odio-api or before this process
// and gets watched from here; one failed is shown as such. Nothing here
// starts the unit.
func (s *Services) UpgradeState(report *upgrade.Report) (running bool, note actionNote) {
	s.mu.Lock()
	watching, note := s.watching, s.upgradeNote
	s.mu.Unlock()
	if watching {
		return true, actionNote{}
	}
	if report == nil || !report.UpgradeAvailable {
		return false, note
	}
	u, err := s.showUpgrade()
	switch {
	case err != nil:
		return false, note
	case u.running():
		s.log.Printf("upgrade: %s is activating, watching it", UpgradeUnit)
		s.watchUpgrade()
		return true, actionNote{}
	case u.Active == "failed":
		return false, u.note()
	}
	return false, note
}

// SetDAC escalates through `sudo -n odioctl dac set <id>`.
func (s *Services) SetDAC(id string) (string, error) {
	if _, ok := dac.ByID(id); !ok {
		return "", userErrorf("unknown DAC id %q", id)
	}
	// Plain HTML cannot grey the Apply button out, so re-applying the current
	// selection is one click away: recognise the no-op here rather than
	// escalate through sudo and claim a reboot that nothing needs. Only when
	// odioctl owns the block, though — same id over an unmanaged config.txt
	// does change the file (takes ownership, comments the stray lines out).
	if d := s.DacStatus(); d.Managed && d.Current == id {
		return "DAC already set to " + id + " — nothing to apply.", nil
	}
	if err := s.runDac("dac", "set", id); err != nil {
		return "", err
	}
	s.changed()
	return "DAC set to " + id + " — reboot required.", nil
}

// Reboot asks logind, as the target user: odios' polkit rule lets that user
// reboot, no sudo. The flag under /run goes with the boot.
func (s *Services) Reboot() (string, error) {
	if err := runChecked(s.run.User, []string{"systemctl", "reboot"}, "systemctl reboot"); err != nil {
		return "", err
	}
	s.log.Printf("reboot requested")
	return "Rebooting — the box is back in a minute.", nil
}

// UnsetDAC removes the odioctl block from config.txt, through sudo.
func (s *Services) UnsetDAC() (string, error) {
	if err := s.runDac("dac", "unset"); err != nil {
		return "", err
	}
	s.changed()
	return "DAC block removed — reboot required.", nil
}

func (s *Services) runDac(args ...string) error {
	if s.cfg.ConfigTxt != "" {
		args = append(args, "--config", s.cfg.ConfigTxt)
	}
	return runChecked(s.run.Privileged, args, "odioctl dac")
}

// runChecked turns any subprocess failure into a UserError banner.
func runChecked(run RunFn, args []string, what string) error {
	res, err := run(args)
	if err != nil {
		return userErrorf("cannot run %s: %v", what, err)
	}
	if res.Code != 0 {
		detail := strings.TrimSpace(res.Stderr)
		if detail == "" {
			detail = strings.TrimSpace(res.Stdout)
		}
		if detail == "" {
			detail = fmt.Sprintf("exit %d", res.Code)
		}
		return userErrorf("%s failed: %s", what, detail)
	}
	return nil
}

func stateErrorMsg(path string, err error) string {
	if os.IsNotExist(err) {
		return path + " not found"
	}
	return fmt.Sprintf("cannot read state.json: %v", err)
}
